### PR TITLE
[SYCL] Make host device and host platform always available, even when forcin…

### PR DESCRIPTION
**Pull request by jbrodman**
…g a device type with SYCL_DEVICE_TYPE

The current behavior violates the SYCL Specification.  The host device/platform should always be available.

int main() { 
  device myHost(host_selector{});
  return 0;
}

That fails when run with SYCL_DEVICE_TYPE=CPU/GPU/ACC and it shouldn't.

Signed-off-by: James Brodman <james.brodman@intel.com>

### CODE REVIEW

File: sycl/source/device.cpp [hunk 1]
```diff
@@ -41,10 +41,20 @@ vector_class<device> device::get_devices(info::device_type deviceType) {
   if (detail::match_types(deviceType, forced_type)) {
     detail::force_type(deviceType, forced_type);
     for (const auto &plt : platform::get_platforms()) {
-      vector_class<device> found_devices(plt.get_devices(deviceType));
-      if (!found_devices.empty())
-        devices.insert(devices.end(), found_devices.begin(),
-                       found_devices.end());
+      // Host device must always be available.
```
Comment (vladimirlaz) [inline sycl/source/device.cpp]: We can left code like it was before if platform_impl::get_devices() is changed to return host 
device for host platform. To do that you can replace
sycl/source/detail/platform_impl.cpp (lines 196-204)

if (is_host()) 
   return Res.resize(1); // default device constructor creates host device

Comment (jbrodman) [reply sycl/source/device.cpp]: I didn't do that b/c get_devices(something_not_host) is supposed to return nothing in that case.

**Comment by jbrodman** [comment]
> Please add LIT test following example in the PR description.

Added a test for host device and another for host platform.

File: sycl/test/basic_tests/host_always_works.cpp [hunk 2]
```diff
@@ -0,0 +1,23 @@
+// RUN: %clangxx -fsycl %s -o %t1.out
+// RUN: env SYCL_DEVICE_TYPE=HOST %t1.out
+// RUN: %CPU_RUN_PLACEHOLDER %t1.out
```
Comment (vladimirlaz) [inline sycl/test/basic_tests/host_always_works.cpp]: Please add the following line to cover accelerator as well:
// RUN: %ACC_RUN_PLACEHOLDER %t1.out

File: sycl/test/basic_tests/host_platform_avail.cpp [hunk 3]
```diff
@@ -0,0 +1,41 @@
+// RUN: %clangxx -fsycl %s -o %t1.out
+// RUN: env SYCL_DEVICE_TYPE=HOST %t1.out
+// RUN: %CPU_RUN_PLACEHOLDER %t1.out
+// RUN: %GPU_RUN_PLACEHOLDER %t1.out
```
Comment (vladimirlaz) [inline sycl/test/basic_tests/host_platform_avail.cpp]: Please add the following line to cover accelerator as well:
// RUN: %CPU_RUN_PLACEHOLDER %t1.out

File: sycl/test/basic_tests/host_always_works.cpp (hunk 2 above)
Comment (jbrodman) [reply sycl/test/basic_tests/host_always_works.cpp]: Done.

File: sycl/test/basic_tests/host_platform_avail.cpp (hunk 3 above)
Comment (jbrodman) [reply sycl/test/basic_tests/host_platform_avail.cpp]: Done.

File: sycl/source/device.cpp [hunk 4]
```diff
@@ -41,10 +41,20 @@ vector_class<device> device::get_devices(info::device_type deviceType) {
...
+      if (plt.is_host()) {
+        vector_class<device> host_device(
+            plt.get_devices(info::device_type::host));
+        if (!host_device.empty())
+          devices.insert(devices.end(), host_device.begin(),
+                         host_device.end());
+      }
+      else {
```
Comment (bader) [inline sycl/source/device.cpp]: Please, apply clang-format to fix formatting.

File: sycl/source/device.cpp [hunk 5]
```diff
@@ -41,10 +41,20 @@ vector_class<device> device::get_devices(info::device_type deviceType) {
...
     for (const auto &plt : platform::get_platforms()) {
-      vector_class<device> found_devices(plt.get_devices(deviceType));
-      if (!found_devices.empty())
-        devices.insert(devices.end(), found_devices.begin(),
-                       found_devices.end());
+      // Host device must always be available.
+      if (plt.is_host()) {
+        vector_class<device> host_device(
+            plt.get_devices(info::device_type::host));
+        if (!host_device.empty())
```
Comment (romanovvlad) [inline sycl/source/device.cpp]: Isn't this if redundant? I think insert will correctly handle case when begin iterator == end iterator.

File: sycl/test/basic_tests/host_platform_avail.cpp [hunk 6]
```diff
@@ -0,0 +1,42 @@
...
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+
+#include <CL/sycl.hpp>
+
+using namespace cl::sycl;
+
+class foo;
```
Comment (romanovvlad) [inline sycl/test/basic_tests/host_platform_avail.cpp]: Why do you need class foo?

File: sycl/test/basic_tests/host_platform_avail.cpp [hunk 7]
```diff
@@ -0,0 +1,42 @@
...
+int main() {
+  queue q;
+  auto dev = q.get_device();
+  auto dev_platform = dev.get_platform();
+
```
Comment (romanovvlad) [inline sycl/test/basic_tests/host_platform_avail.cpp]: Why do you need that? I think the body of the if on line 25 should be enough for the test.

File: sycl/source/device.cpp (hunk 5 above)
Comment (jbrodman) [reply sycl/source/device.cpp]: No - get_devices is supposed to only return devices of the passed type, hence the special casing for host devices.

File: sycl/test/basic_tests/host_platform_avail.cpp (hunk 6 above)
Comment (jbrodman) [reply sycl/test/basic_tests/host_platform_avail.cpp]: Copy/paste artifact.

File: sycl/test/basic_tests/host_platform_avail.cpp (hunk 7 above)
Comment (jbrodman) [reply sycl/test/basic_tests/host_platform_avail.cpp]: Good point.

File: sycl/source/device.cpp (hunk 4 above)
Comment (jbrodman) [reply sycl/source/device.cpp]: Done.

**Comment by jbrodman** [comment]
> LGTM. Thanks!

👍
Good to merge? This is gating another priority PR I need to file.

**Pull request merged**